File: ios/Sources/CompanionCore/ChatPreferences.swift
// The parts of the chat's preferences that are logic rather than screen:
// how much of a bot's working-out the transcript shows, and what sits on
// the composer's chip row.
//
// These live in the core, away from SwiftUI, because they are the pieces
// worth testing — the views that read them are wiring.
import Foundation

/// How much of a bot's activity the transcript shows.
public enum ActivityDetail: String, CaseIterable, Codable, Sendable {
    /// Every chip, as the harness sent them.
    case full
    /// Consecutive chips fold into one summary; failures never fold.
    case reduced
    /// No activity chips at all.
    case hidden

    public var label: String {
        switch self {
        case .full: "Full"
        case .reduced: "Reduced"
        case .hidden: "Hidden"
        }
    }

    public var caption: String {
        switch self {
        case .full: "Every step a bot takes."
        case .reduced: "Steps fold into one line. Failures always show."
        case .hidden: "No activity, only messages."
        }
    }
}

/// When the island plays a bot's face on opening a chat.
public enum IslandIntro: String, CaseIterable, Codable, Sendable {
    case always
    /// The first time each bot is opened, then never again for that bot.
    case oncePerBot
    case never

    public var label: String {
        switch self {
        case .always: "Always"
        case .oncePerBot: "First time per bot"
        case .never: "Never"
        }
    }
}

/// One chip on the composer's quick-reply row.
public struct QuickReply: Codable, Hashable, Identifiable, Sendable {
    public var id: String
    public var title: String
    public var prompt: String
    public var icon: String

    public init(id: String = UUID().uuidString, title: String, prompt: String, icon: String) {
        self.id = id
        self.title = title
        self.prompt = prompt
        self.icon = icon
    }

    /// The four the composer shipped with, kept as the reset target.
    public static let defaults: [QuickReply] = [
        QuickReply(id: "default.diff", title: "Show diff", prompt: "Show latest git diff", icon: "arrow.triangle.pull"),
        QuickReply(id: "default.tests", title: "Run tests", prompt: "Run all automated tests", icon: "checkmark.seal"),
        QuickReply(id: "default.explain", title: "Explain steps", prompt: "Explain the changes in detail", icon: "text.bubble"),
        QuickReply(id: "default.next", title: "What's next?", prompt: "What should we do next?", icon: "sparkles"),
    ]

    /// Icons offered by the editor. Small on purpose: a full symbol browser
    /// is a different feature, and twelve covers what a chip is ever for.
    public static let iconChoices = [
        "sparkles", "arrow.triangle.pull", "checkmark.seal", "text.bubble",
        "hammer", "ladybug", "doc.text", "terminal",
        "paperplane", "magnifyingglass", "clock.arrow.circlepath", "list.bullet",
    ]

    public static func encode(_ replies: [QuickReply]) -> String {
        guard let data = try? JSONEncoder().encode(replies) else { return "" }
        return String(decoding: data, as: UTF8.self)
    }

    /// Anything unreadable becomes the defaults. An empty *string* is a
    /// store that has never been written; an empty *list* is a row the user
    /// deliberately cleared, and those must not mean the same thing.
    public static func decode(_ json: String) -> [QuickReply] {
        guard !json.isEmpty, let data = json.data(using: .utf8) else { return defaults }
        guard let decoded = try? JSONDecoder().decode([QuickReply].self, from: data) else { return defaults }
        let ids = decoded.map { $0.id.trimmingCharacters(in: .whitespacesAndNewlines) }
        guard ids.allSatisfy({ !$0.isEmpty }), Set(ids).count == ids.count else { return defaults }
        return decoded
    }
}

/// A completed turn's intermediate replies. The terminal answer stays outside.
public struct AssistantTurnFold: Hashable, Sendable {
    public let turnId: String
    public let messages: [Message]
    public let elapsed: Double

    public var label: String {
        guard elapsed >= 1_000 else { return "Worked" }
        let seconds = Int(elapsed / 1_000)
        let duration = seconds < 60 ? "\(seconds)s" : "\(seconds / 60)m \(String(format: "%02d", seconds % 60))s"
        return "Worked for \(duration)"
    }
}

/// A transcript entry: a message, a run of tools, or completed narration.
public enum TranscriptRow: Identifiable, Hashable, Sendable {
    case message(Message)
    case activityRun([Message])
    case assistantTurn(AssistantTurnFold)

    /// The message a row answers for — its first, which is the one whose
    /// time and sender the transcript reads.
    public var head: Message {
        switch self {
        case let .message(message): message
        case let .activityRun(items): items[0]
        case let .assistantTurn(turn): turn.messages[0]
        }
    }

    public var id: String {
        switch self {
        case let .message(message): message.id
        case let .activityRun(items): "run.\(items[0].id)"
        case let .assistantTurn(turn): "turn.\(turn.turnId)"
        }
    }

    public var at: Double { head.at }
    /// The last timestamp covered by this row. Date separators compare the
    /// next row with this value so a folded run cannot manufacture a gap.
    public var endAt: Double {
        switch self {
        case let .message(message): message.at
        case let .activityRun(items): items.last?.at ?? head.at
        case let .assistantTurn(turn): turn.messages.last?.at ?? head.at
        }
    }
    public var role: Message.Role { head.role }
    public var kind: Message.Kind { head.kind }
    public var senderName: String? { head.from?.name }
}

/// The one line a roster row shows under a chat's name.
///
/// Folded by the same rule as the transcript, and for the same reason: a
/// reader who has turned activity off has said they do not want to see tool
/// calls, and the roster is where they see the most of them — one per chat,
/// on the screen they spend the most time on. Reading the preview off the
/// raw last message made "Hidden" mean "hidden in one place".
public func rosterPreview(_ messages: [Message], detail: ActivityDetail) -> String {
    guard let last = transcriptRows(messages, detail: detail).last else { return "" }
    switch last {
    case let .message(message):
        return previewText(of: message)
    case let .activityRun(items):
        let running = items.contains { $0.tool?.ok == nil }
        return "\(running ? "Running" : "Ran") \(items.count) steps"
    case let .assistantTurn(turn):
        return turn.label
    }
}

/// What a single message reads as in a roster row.
func previewText(of message: Message) -> String {
    switch message.kind {
    case .text: return message.webhookContent?.task ?? message.text ?? ""
    // a pending card's question is the preview; the roster row already
    // says "waiting on you" beside it
    case .options:
        guard let card = message.card else { return "" }
        return card.isPending && !card.subtitle.isEmpty ? card.subtitle : card.title
    case .secret:
        return message.secret?.label ?? message.text ?? "Credential required"
    case .activity: return message.tool?.name ?? ""
    case .screen: return "Screenshot"
    case .digest: return ""
    case .compaction: return message.compaction?.chipText ?? message.text ?? ""
    case .unknown: return message.text ?? ""
    }
}

/// Rows the harness writes about a turn rather than in it: tool chips and,
/// since Phase 0, the digest and compaction receipts. Hidden together,
/// because a reader who turned activity off does not want the summary of
/// exactly those calls either.
public func isActivityReceipt(_ message: Message) -> Bool {
    switch message.kind {
    case .activity, .digest, .compaction: return true
    default: return false
    }
}

/// Folds a transcript to the requested level of detail.
///
/// A failed step is never folded away: the reason to turn activity down is
/// the successful noise, and losing the one chip that says something went
/// wrong would make `reduced` a worse default than `full`.
public func transcriptRows(_ messages: [Message], detail: ActivityDetail) -> [TranscriptRow] {
    let messages = messages.filter { $0.kind != .digest }
    // Fold only explicitly completed turns; never guess that the last reply
    // is final on an older server or while the bot is still working.
    var narration: [String: [Message]] = [:]
    var startedAt: [String: Double] = [:]
    var lastUserAt: Double?
    var folds: [String: AssistantTurnFold] = [:]
    var hiddenIDs = Set<String>()
    for message in messages {
        if message.role == .user { lastUserAt = message.at }
        guard message.role == .bot, message.kind == .text,
              let turnID = message.turnId, !turnID.isEmpty else { continue }
        if message.turnTerminal == true {
            guard let items = narration.removeValue(forKey: turnID), !items.isEmpty else { continue }
            folds[items[0].id] = AssistantTurnFold(
                turnId: turnID, messages: items,
                elapsed: max(0, message.at - (startedAt[turnID] ?? items[0].at))
            )
            hiddenIDs.formUnion(items.map(\.id))
        } else {
            if narration[turnID] == nil { startedAt[turnID] = lastUserAt ?? message.at }
            narration[turnID, default: []].append(message)
        }
    }

    var rows: [TranscriptRow] = []
    var run: [Message] = []

    // A run of one is not worth a summary — emit the chip itself.
    func flush() {
        if run.count > 1 {
            rows.append(.activityRun(run))
        } else {
            rows.append(contentsOf: run.map(TranscriptRow.message))
        }
        run.removeAll()
    }

    for message in messages {
        if let turn = folds[message.id] {
            flush()
            rows.append(.assistantTurn(turn))
            continue
        }
        if hiddenIDs.contains(message.id) { continue }
        if detail == .hidden && isActivityReceipt(message) { continue }
        if detail != .reduced {
            rows.append(.message(message))
            continue
        }
        guard message.kind == .activity else {
            flush()
            rows.append(.message(message))
            continue
        }
        if message.tool?.ok == false {
            flush()
            rows.append(.message(message))
            continue
        }
        run.append(message)
    }
    flush()
    return rows
}

File: android/core/src/main/kotlin/com/openmausbot/companion/core/ChatPreferences.kt
package com.openmausbot.companion.core

import java.util.UUID
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString

/** How much of a bot's working activity the transcript shows. */
enum class ActivityDetail(val wireValue: String, val label: String, val caption: String) {
    FULL("full", "Full", "Every step a bot takes."),
    REDUCED("reduced", "Reduced", "Steps fold into one line. Failures always show."),
    HIDDEN("hidden", "Hidden", "No activity, only messages."),
    ;

    companion object {
        fun fromWire(value: String?): ActivityDetail =
            entries.firstOrNull { it.wireValue == value } ?: FULL
    }
}

/** One user-editable chip on the composer's quick-reply row. */
@Serializable
data class QuickReply(
    val id: String = UUID.randomUUID().toString(),
    val title: String,
    val prompt: String,
    val icon: String,
) {
    companion object {
        val DEFAULTS: List<QuickReply> = listOf(
            QuickReply("default.diff", "Show diff", "Show latest git diff", "diff"),
            QuickReply("default.tests", "Run tests", "Run all automated tests", "tests"),
            QuickReply("default.explain", "Explain steps", "Explain the changes in detail", "explain"),
            QuickReply("default.next", "What's next?", "What should we do next?", "next"),
        )

        val ICON_CHOICES: List<String> = listOf(
            "next", "diff", "tests", "explain", "build", "bug", "document", "terminal",
            "send", "search", "history", "list",
        )

        fun encode(replies: List<QuickReply>): String =
            runCatching { CompanionJson.encodeToString(replies) }.getOrDefault("")

        /**
         * An empty or corrupt store falls back to defaults. An encoded empty list is a deliberate
         * choice and remains empty.
         */
        fun decode(json: String): List<QuickReply> {
            if (json.isEmpty()) return DEFAULTS
            val decoded = runCatching {
                CompanionJson.decodeFromString<List<QuickReply>>(json)
            }.getOrElse { return DEFAULTS }
            val ids = decoded.map { it.id.trim() }
            if (ids.any(String::isEmpty) || ids.toSet().size != ids.size) return DEFAULTS
            return decoded
        }
    }
}

/** A transcript item: one message, consecutive activity, or completed narration. */
sealed interface TranscriptRow {
    val head: Message
    val id: String
    val at: Double get() = head.at
    val endAt: Double
    val role: Message.Role get() = head.role
    val kind: Message.Kind get() = head.kind
    val senderName: String? get() = head.from?.name

    /** Search and pagination anchors can land inside a folded row. */
    fun containsMessage(messageId: String): Boolean = when (this) {
        is Single -> message.id == messageId
        is ActivityRun -> items.any { it.id == messageId }
        is AssistantTurn -> items.any { it.id == messageId }
    }

    data class Single(val message: Message) : TranscriptRow {
        override val head: Message get() = message
        override val id: String get() = message.id
        override val endAt: Double get() = message.at
    }

    data class ActivityRun(val items: List<Message>) : TranscriptRow {
        init {
            require(items.isNotEmpty()) { "An activity run must contain at least one message." }
        }

        override val head: Message get() = items.first()
        override val id: String get() = "run.${head.id}"
        override val endAt: Double get() = items.last().at
        val running: Boolean get() = items.any { it.tool?.ok == null }
    }

    data class AssistantTurn(val turnId: String, val items: List<Message>, val elapsed: Double) : TranscriptRow {
        init {
            require(items.isNotEmpty()) { "A turn fold must contain narration." }
        }

        override val head: Message get() = items.first()
        override val id: String get() = "turn.$turnId"
        override val endAt: Double get() = items.last().at
        val label: String get() {
            if (elapsed < 1_000) return "Worked"
            val seconds = (elapsed / 1_000).toLong()
            val duration = if (seconds < 60) "${seconds}s"
                else "${seconds / 60}m ${(seconds % 60).toString().padStart(2, '0')}s"
            return "Worked for $duration"
        }
    }
}

/**
 * The one line a roster row shows under a chat's name.
 *
 * Folded by the same rule as the transcript, and for the same reason: a reader who has turned
 * activity off has said they do not want to see tool calls, and the roster is where they see the
 * most of them — one per chat, on the screen they spend the most time on. Reading the preview off
 * the raw last message made "Hidden" mean "hidden in one place".
 */
fun rosterPreview(messages: List<Message>, detail: ActivityDetail): String =
    when (val last = transcriptRows(messages, detail).lastOrNull()) {
        null -> ""
        is TranscriptRow.Single -> previewText(last.message)
        is TranscriptRow.ActivityRun ->
            "${if (last.running) "Running" else "Ran"} ${last.items.size} steps"
        is TranscriptRow.AssistantTurn -> last.label
    }

/** What a single message reads as in a roster row. */
internal fun previewText(message: Message): String = when (message.kind) {
    Message.Kind.TEXT -> message.webhookContent?.task ?: message.text.orEmpty()
    // a pending card's question is the preview; the roster row already says
    // "waiting on you" beside it
    Message.Kind.OPTIONS -> {
        val card = message.card
        when {
            card == null -> ""
            card.isPending && card.subtitle.isNotEmpty() -> card.subtitle
            else -> card.title
        }
    }
    Message.Kind.ACTIVITY -> message.tool?.name.orEmpty()
    Message.Kind.SCREEN -> "Screenshot"
    Message.Kind.DIGEST -> ""
    Message.Kind.COMPACTION -> message.compaction?.chipText ?: message.text.orEmpty()
    Message.Kind.UNKNOWN -> message.text.orEmpty()
}

/**
 * Rows the harness writes about a turn rather than in it: tool chips and, since
 * Phase 0, the digest and compaction receipts. Hidden together, because a reader
 * who turned activity off does not want the summary of exactly those calls either.
 * Port of `isActivityReceipt` in `ChatPreferences.swift`.
 */
fun isActivityReceipt(message: Message): Boolean = when (message.kind) {
    Message.Kind.ACTIVITY, Message.Kind.DIGEST, Message.Kind.COMPACTION -> true
    else -> false
}

/**
 * Fold a transcript to the selected activity detail. Failed steps are never folded in reduced
 * mode; hidden mode intentionally removes all activity, including failures.
 */
fun transcriptRows(messages: List<Message>, detail: ActivityDetail): List<TranscriptRow> {
    // Only a server completion marker makes narration foldable. Legacy and
    // unfinished turns stay visible, matching desktop and iOS.
    val narration = mutableMapOf<String, MutableList<Message>>()
    val startedAt = mutableMapOf<String, Double>()
    var lastUserAt: Double? = null
    val folds = mutableMapOf<String, TranscriptRow.AssistantTurn>()
    val hiddenIds = mutableSetOf<String>()
    for (message in messages) {
        if (message.role == Message.Role.USER) lastUserAt = message.at
        val turnId = message.turnId?.takeIf { it.isNotEmpty() } ?: continue
        if (message.role != Message.Role.BOT || message.kind != Message.Kind.TEXT) continue
        if (message.turnTerminal == true) {
            val items = narration.remove(turnId)?.takeIf { it.isNotEmpty() } ?: continue
            folds[items.first().id] = TranscriptRow.AssistantTurn(
                turnId, items.toList(), (message.at - (startedAt[turnId] ?: items.first().at)).coerceAtLeast(0.0),
            )
            hiddenIds.addAll(items.map { it.id })
        } else {
            if (turnId !in narration) startedAt[turnId] = lastUserAt ?: message.at
            narration.getOrPut(turnId) { mutableListOf() }.add(message)
        }
    }
    return buildList {
        val run = mutableListOf<Message>()
        fun flush() {
            when (run.size) {
                0 -> Unit
                1 -> add(TranscriptRow.Single(run.single()))
                else -> add(TranscriptRow.ActivityRun(run.toList()))
            }
            run.clear()
        }

        messages.forEach { message ->
            if (message.kind == Message.Kind.DIGEST) return@forEach
            val turn = folds[message.id]
            if (turn != null) {
                flush()
                add(turn)
            } else if (message.id in hiddenIds || (detail == ActivityDetail.HIDDEN && isActivityReceipt(message))) {
                // The reversible turn fold owns narration; Hidden owns tools.
            } else if (detail != ActivityDetail.REDUCED || message.kind != Message.Kind.ACTIVITY) {
                flush()
                add(TranscriptRow.Single(message))
            } else if (message.tool?.ok == false) {
                flush()
                add(TranscriptRow.Single(message))
            } else {
                run += message
            }
        }
        flush()
    }
}
